src/sentry/api/serializers/models/release.py

```diff
@@ -260,16 +260,27 @@ def __get_release_data_no_environment(self, project, item_list):
                 release__in=item_list, new_groups__isnull=False
             ).values_list("project_id", "release_id", "new_groups"):
                 group_counts_by_release.setdefault(release_id, {})[project_id] = new_groups
+
         return first_seen, last_seen, group_counts_by_release
 
-    def __get_release_data_with_environments(self, project, item_list, environments):
-        release_project_envs = ReleaseProjectEnvironment.objects.filter(
-            release__in=item_list, environment__name__in=environments
-        ).select_related("release")
-        if project is not None:
-            release_project_envs = release_project_envs.filter(project=project)
+    def _get_release_adoption_stages(self, release_project_envs):
+        adoption_stages = defaultdict(dict)
+
+        for release_project_env in release_project_envs:
+            if (
+                release_project_env.project.slug
+                not in adoption_stages[release_project_env.release.version]
+            ):
+                adoption_stages[release_project_env.release.version][
+                    release_project_env.project.slug
+                ] = release_project_env.adoption_string
```

Comment:
You can simplify this a little via:
`adoption_stages[release_project_env.release.version].setdefault(release_project_env.release.version, release_project_env.adoption_string)`

How does this work if the release has a project that has multiple different stages due to environments?

Reply:
It would just select the first environment that we get from sorting by `-first_seen`. I think we might end up selecting based on the environment with the highest number of sessions in the near future

Reply:
I wonder if we should be aggregating them or something... like 
- unadopted means no environment has adopted or replaced
- adopted means at least one environment has adopted
- replaced means all environments are unadopted or replaced, and at least one environment is replaced.

Reply:
I'll bring this up with the group and see what they think about it as an alternative selection method